app/celer-g4/PGPrimaryGeneratorAction.cc

```diff
@@ -57,32 +55,38 @@ PGPrimaryGeneratorAction::PGPrimaryGeneratorAction(
 void PGPrimaryGeneratorAction::GeneratePrimaries(G4Event* event)
 {
     CELER_EXPECT(event);
+    CELER_EXPECT(event->GetEventID() >= 0);
 
-    if (event_count_ == num_events_)
+    size_type event_id = event->GetEventID();
+    if (event_id >= num_events_)
     {
         return;
     }
 
+    // Seed with an independent value for each event. Since Geant4 schedules
+    // events dynamically, the same event ID may not be mapped to the same
+    // thread across multiple runs. For reproducibility, Geant4 reseeds each
+    // worker thread's RNG at the start of each event with a seed calculated
+    // from the event ID.
+    rng_.seed(seed_ + event_id);
```

Comment:
Currently the primary generator RNGs are all `std::mt1997`, but we should consider some sort of wrapper interface that lets us do seed+stride for the XORWOW generators (and any others we implement in the future)